masonry/src/tree_arena.rs

```diff
@@ -0,0 +1,297 @@
+// Copyright 2024 the Xilem Authors
+// SPDX-License-Identifier: Apache-2.0
+
+// TODO - Factor out into a separate crate.
+#![allow(dead_code)]
+
+struct TreeNode<Item> {
+    id: u64,
+    item: Item,
+    children: Vec<TreeNode<Item>>,
+}
+
+#[derive(Default)]
+pub struct TreeArena<Item> {
+    roots: Vec<TreeNode<Item>>,
+}
+
+pub struct TreeArenaToken<'a, Item> {
+    children: &'a Vec<TreeNode<Item>>,
+}
+
+impl<'a, Item> Clone for TreeArenaToken<'a, Item> {
+    fn clone(&self) -> Self {
+        *self
+    }
+}
+
+impl<'a, Item> Copy for TreeArenaToken<'a, Item> {}
+
+pub struct TreeArenaTokenMut<'a, Item> {
+    children: &'a mut Vec<TreeNode<Item>>,
+}
+
+#[cfg(FALSE)]
+struct TreeArena<Item> {
+    items: HashMap<u64, UnsafeCell<Item>>,
+    parents: HashMap<u64, u64>,
+}
+
+#[cfg(FALSE)]
+struct TreeArenaToken<'a, Item> {
+    arena: &'a TreeArena<Item>,
+    id: u64,
+}
+
+#[cfg(FALSE)]
+struct TreeArenaTokenMut<'a, Item> {
+    arena: &'a TreeArena<Item>,
+    id: u64,
+}
+
+impl<Item> TreeArena<Item> {
+    pub fn new() -> Self {
+        TreeArena { roots: Vec::new() }
+    }
+
+    pub fn root_token(&self) -> TreeArenaToken<'_, Item> {
+        TreeArenaToken {
+            children: &self.roots,
+        }
+    }
+
+    pub fn root_token_mut(&mut self) -> TreeArenaTokenMut<'_, Item> {
+        TreeArenaTokenMut {
+            children: &mut self.roots,
+        }
+    }
+
+    // TODO - Use list of parents
+    pub fn find(&self, id: u64) -> Option<(&Item, TreeArenaToken<'_, Item>)> {
```

Comment:
This should have a comment saying either:
1) This is `O(n)` where `n` is the total number of items in the subtree; OR
2) This is currently as in (1), but will be `O(1)` in future.